File: 2023/kotlin-aoc-2023/src/test/kotlin/day01/Day01.kt
package day01

import FileUtil.readInputFileToList
import io.kotest.matchers.shouldBe
import org.junit.jupiter.api.Test

class Day01 {
    @Test
    internal fun `part 1 sample input`() {
        day1Part1SumFirstAndLastDigits(testInput) shouldBe 142
    }

    @Test
    internal fun `part 1 real input`() {
        day1Part1SumFirstAndLastDigits(readInputFileToList("day01.txt")) shouldBe 54951
    }

    @Test
    internal fun `part 2 sample input`() {
        day1Part2SumFirstAndLastDigitsAfterSubstitutingDigitsForWords(testInput2) shouldBe 281
    }

    @Test
    internal fun `part 2 real input`() {
        day1Part2SumFirstAndLastDigitsAfterSubstitutingDigitsForWords(readInputFileToList("day01.txt")) shouldBe 55218
    }

    @Test
    fun `correctly maps overlapping numbers`() {
        "ppkeightwo5ggthreet".substituteDigitsForWords() shouldBe "ppk825gg3t"
        "eightwothree".substituteDigitsForWords() shouldBe "823"
    }
}

// What a massive hack!
private fun String.substituteDigitsForWords() =
    replace("oneight", "18")
        .replace("twone", "21")
        .replace("threeight", "38")
        .replace("fiveeight", "58")
        .replace("sevenine", "79")
        .replace("eightwo", "82")
        .replace("eighthree", "83")
        .replace("nineight", "98")
        .replace("one", "1")
        .replace("two", "2")
        .replace("three", "3")
        .replace("four", "4")
        .replace("five", "5")
        .replace("six", "6")
        .replace("seven", "7")
        .replace("eight", "8")
        .replace("nine", "9")

private fun day1Part1SumFirstAndLastDigits(input: List<String>) =
    input.map { line ->
        line.concatenateFirstAndLastDigitsAndConvertToInt()
    }.sum()

private fun day1Part2SumFirstAndLastDigitsAfterSubstitutingDigitsForWords(input: List<String>) =
    input.map {
        it.substituteDigitsForWords().concatenateFirstAndLastDigitsAndConvertToInt()
    }.sum()

private fun String.concatenateFirstAndLastDigitsAndConvertToInt(): Int {
    val digits = filter { it.isDigit() }
    return "${digits.first()}${digits.last()}".toInt()
}

val testInput =
    """
        1abc2
        pqr3stu8vwx
        a1b2c3d4e5f
        treb7uchet
    """.trimIndent().lines()

val testInput2 =
    """
        two1nine
        eightwothree
        abcone2threexyz
        xtwone3four
        4nineeightseven2
        zoneight234
        7pqrstsixteen
    """.trimIndent().lines()

File: 2023/clj-aoc-2023/src/clj_aoc_2023/day01.clj
(ns clj-aoc-2023.day01)

(defn to-digit [it]
  (cond
    (re-matches #"\d" it) (Integer. it)
    (= "one" it) 1
    (= "two" it) 2
    (= "three" it) 3
    (= "four" it) 4
    (= "five" it) 5
    (= "six" it) 6
    (= "seven" it) 7
    (= "eight" it) 8
    (= "nine" it) 9))

(defn first-digit-or-word [line]
  (->> line
       (re-seq #"(?=([0-9]|one|two|three|four|five|six|seven|eight|nine))")
       (first)
       (second)
       (to-digit)))

(defn last-digit-or-word [line]
  (->> line
       (re-seq #"(?=(\d|one|two|three|four|five|six|seven|eight|nine))")
       (last)
       (second)
       (to-digit)))

(defn day01-part2 [input]
  (->> input
       (map #(+
              (* 10 (first-digit-or-word %))
              (last-digit-or-word %)))
       (reduce + 0)))

(defn first-digit [line]
  (->> line
       (re-seq #"\d")
       (first)
       (Integer.)))

(defn last-digit [line]
  (->> line
       (re-seq #"\d")
       (last)
       (Integer.)))

(defn value-of-first-and-last [line]
  (+ (* (first-digit line) 10)
     (last-digit line)))

(defn day01-part1 [input]
  (->> input
       (map value-of-first-and-last)
       (reduce + 0)))
